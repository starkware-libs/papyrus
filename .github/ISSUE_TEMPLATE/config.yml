---
blank_issues_enabled: false
contact_links:
  - name: Papyrus Community Support
    url: https://github.com/starkware-libs/discussions
    about: Please ask and answer questions here.
