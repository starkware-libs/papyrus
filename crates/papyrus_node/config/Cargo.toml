[package]
name = "config"
version = "0.1.0"
edition = "2021"

[dependencies]
anyhow = { version = "1.0.44" }
ron = { version ="0.7" }
serde = { version = "1.0.130", features = ["derive"] }
