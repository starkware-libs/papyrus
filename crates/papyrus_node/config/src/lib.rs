#[cfg(test)]
mod config_test;

use std::fs;
use std::time::Duration;

use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize)]
pub struct GatewayConfig {
    pub server_ip: String,
}

#[derive(Serialize, Deserialize)]
pub struct MonitoringGatewayConfig {
    pub server_ip: String,
}

#[derive(Serialize, Deserialize)]
pub struct StorageConfig {
    pub db_config: DbConfig,
}

#[derive(Serialize, Deserialize)]
pub struct DbConfig {
    pub path: String,
    pub max_size: usize,
}

#[derive(Serialize, Deserialize)]
pub struct CentralSourceConfig {
    pub url: String,
    pub retry_config: RetryConfig,
}

/// A configuration for the retry mechanism.
#[derive(Serialize, Deserialize)]
pub struct RetryConfig {
    /// The initial waiting time in milliseconds.
    pub retry_base_millis: u64,
    /// The maximum waiting time in milliseconds.
    pub retry_max_delay_millis: u64,
    /// The maximum number of retries.
    pub max_retries: usize,
}

#[derive(Serialize, Deserialize)]
pub struct SyncConfig {
    pub block_propagation_sleep_duration: Duration,
}

#[derive(Deserialize, Serialize)]
pub struct Config {
    pub central: CentralSourceConfig,
    pub gateway: GatewayConfig,
    pub monitoring_gateway: MonitoringGatewayConfig,
    pub storage: StorageConfig,
    pub sync: SyncConfig,
}

pub fn load_config(path: &'static str) -> anyhow::Result<Config> {
    let config_contents = fs::read_to_string(path).expect("Something went wrong reading the file");
    let config: Config = ron::from_str(&config_contents)?;
    Ok(config)
}
