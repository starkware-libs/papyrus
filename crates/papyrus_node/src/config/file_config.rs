use std::collections::HashMap;
use std::path::Path;
use std::time::Duration;
use std::{env, fs};

use papyrus_gateway::GatewayConfig;
use papyrus_monitoring_gateway::MonitoringGatewayConfig;
use papyrus_storage::db::DbConfig;
use papyrus_storage::StorageConfig;
use papyrus_sync::{CentralSourceConfig, SyncConfig};
use serde::{Deserialize, Serialize};
use starknet_api::core::ChainId;
use starknet_client::RetryConfig;

use crate::config::{Config, ConfigBuilder};

// Defines the expected structure of the configuration file. All the fields are optional so the user
// doesn't have to specify parameters that he doesn't wish to override (in that case the previous
// value remains).
#[derive(Serialize, Deserialize, PartialEq, Debug)]
#[serde(deny_unknown_fields)]
pub struct FileConfigFormat {
    chain_id: Option<ChainId>,
    central: Option<Central>,
    gateway: Option<Gateway>,
    monitoring_gateway: Option<MonitoringGateway>,
    storage: Option<Storage>,
    sync: Option<Sync>,
}

impl FileConfigFormat {
    // Apply the configuration as given by the file on the configuration builder.
    pub(crate) fn update_config(self, builder: &mut ConfigBuilder) {
        if let Some(chain_id) = self.chain_id {
            builder.chain_id = chain_id;
        }

        if let Some(central) = self.central {
            central.update_central(&mut builder.config.central);
        }

        if let Some(gateway) = self.gateway {
            gateway.update_gateway(&mut builder.config.gateway);
        }

        if let Some(monitoring_gateway) = self.monitoring_gateway {
            monitoring_gateway.update_monitoring_gateway(&mut builder.config.monitoring_gateway);
        }

        if let Some(storage) = self.storage {
            storage.update_storage(&mut builder.config.storage);
        }

        // Sync is optional, override it only if it is Some in the builder.
        if let (Some(builder_config), Some(file_config)) = (builder.config.sync.as_mut(), self.sync)
        {
            file_config.update_sync(builder_config)
        }
    }
}

impl From<Config> for FileConfigFormat {
    fn from(config: Config) -> Self {
        FileConfigFormat {
            chain_id: Some(config.gateway.chain_id.clone()),
            central: Some(Central::from(config.central)),
            gateway: Some(Gateway::from(config.gateway)),
            monitoring_gateway: Some(MonitoringGateway::from(config.monitoring_gateway)),
            storage: Some(Storage::from(config.storage)),
            sync: config.sync.map(Sync::from),
        }
    }
}

impl From<GatewayConfig> for Gateway {
    fn from(config: GatewayConfig) -> Self {
        Gateway {
            server_address: Some(config.server_address),
            max_events_chunk_size: Some(config.max_events_chunk_size),
            max_events_keys: Some(config.max_events_keys),
        }
    }
}

impl From<CentralSourceConfig> for Central {
    fn from(config: CentralSourceConfig) -> Self {
        Central {
            concurrent_requests: Some(config.concurrent_requests),
            url: Some(config.url),
            http_headers: config.http_headers,
            retry: Some(Retry::from(config.retry_config)),
        }
    }
}

impl From<RetryConfig> for Retry {
    fn from(config: RetryConfig) -> Self {
        Retry {
            retry_base_millis: Some(config.retry_base_millis),
            retry_max_delay_millis: Some(config.retry_max_delay_millis),
            max_retries: Some(config.max_retries),
        }
    }
}

impl From<MonitoringGatewayConfig> for MonitoringGateway {
    fn from(config: MonitoringGatewayConfig) -> Self {
        MonitoringGateway { server_address: Some(config.server_address) }
    }
}

impl From<StorageConfig> for Storage {
    fn from(config: StorageConfig) -> Self {
        Storage { db: Some(Db::from(config.db_config)) }
    }
}

impl From<DbConfig> for Db {
    fn from(config: DbConfig) -> Self {
        // Remove the chain_id from the path.
        let mut path = config.path;
        let last_slash_index =
            path.rfind('/').expect("Remove chain_id from the storage file path failed");
        path.truncate(last_slash_index);
        Db { path: Some(path), max_size: Some(config.max_size) }
    }
}

impl From<SyncConfig> for Sync {
    fn from(config: SyncConfig) -> Self {
        Sync {
            block_propagation_sleep_duration_secs: Some(
                config.block_propagation_sleep_duration.as_secs(),
            ),
            recoverable_error_sleep_duration_secs: Some(
                config.recoverable_error_sleep_duration.as_secs(),
            ),
        }
    }
}

#[derive(Serialize, Deserialize, PartialEq, Debug)]
#[serde(deny_unknown_fields)]
struct Central {
    concurrent_requests: Option<usize>,
    url: Option<String>,
    http_headers: Option<HashMap<String, String>>,
    retry: Option<Retry>,
}

impl Central {
    fn update_central(self, config: &mut CentralSourceConfig) {
        if let Some(concurrent_requests) = self.concurrent_requests {
            config.concurrent_requests = concurrent_requests;
        }
        if let Some(url) = self.url {
            config.url = url;
        }
        if let Some(new_headers) = self.http_headers {
            match &mut config.http_headers {
                Some(current_headers) => {
                    current_headers.extend(new_headers);
                }
                None => {
                    config.http_headers = Some(new_headers);
                }
            };
        }
        if let Some(retry) = self.retry {
            retry.update_retry_config(&mut config.retry_config);
        }
    }
}

#[derive(Serialize, Deserialize, PartialEq, Debug)]
#[serde(deny_unknown_fields)]
struct Retry {
    retry_base_millis: Option<u64>,
    retry_max_delay_millis: Option<u64>,
    max_retries: Option<usize>,
}

impl Retry {
    fn update_retry_config(self, config: &mut RetryConfig) {
        if let Some(retry_base_millis) = self.retry_base_millis {
            config.retry_base_millis = retry_base_millis;
        }
        if let Some(retry_max_delay_millis) = self.retry_max_delay_millis {
            config.retry_max_delay_millis = retry_max_delay_millis;
        }
        if let Some(max_retries) = self.max_retries {
            config.max_retries = max_retries;
        }
    }
}

#[derive(Serialize, Deserialize, PartialEq, Debug)]
#[serde(deny_unknown_fields)]
struct Gateway {
    server_address: Option<String>,
    max_events_chunk_size: Option<usize>,
    max_events_keys: Option<usize>,
}

impl Gateway {
    fn update_gateway(self, config: &mut GatewayConfig) {
        if let Some(server_address) = self.server_address {
            config.server_address = server_address;
        }
        if let Some(max_events_chunk_size) = self.max_events_chunk_size {
            config.max_events_chunk_size = max_events_chunk_size;
        }
        if let Some(max_events_keys) = self.max_events_keys {
            config.max_events_keys = max_events_keys;
        }
    }
}

#[derive(Serialize, Deserialize, PartialEq, Debug)]
#[serde(deny_unknown_fields)]
struct MonitoringGateway {
    server_address: Option<String>,
}

impl MonitoringGateway {
    fn update_monitoring_gateway(self, config: &mut MonitoringGatewayConfig) {
        if let Some(server_address) = self.server_address {
            config.server_address = server_address;
        }
    }
}

#[derive(Serialize, Deserialize, PartialEq, Debug)]
#[serde(deny_unknown_fields)]
struct Storage {
    db: Option<Db>,
}

impl Storage {
    fn update_storage(self, config: &mut StorageConfig) {
        if let Some(db) = self.db {
            db.update_db(&mut config.db_config);
        }
    }
}

#[derive(Serialize, Deserialize, PartialEq, Debug)]
#[serde(deny_unknown_fields)]
struct Db {
    path: Option<String>,
    max_size: Option<usize>,
}

impl Db {
    fn update_db(self, config: &mut DbConfig) {
        if let Some(path) = self.path {
            config.path = path;
        }
        if let Some(max_size) = self.max_size {
            config.max_size = max_size;
        }
    }
}

#[derive(Serialize, Deserialize, PartialEq, Debug)]
#[serde(deny_unknown_fields)]
struct Sync {
    block_propagation_sleep_duration_secs: Option<u64>,
    recoverable_error_sleep_duration_secs: Option<u64>,
}

impl Sync {
    fn update_sync(self, config: &mut SyncConfig) {
        if let Some(block_propagation_sleep_duration) = self.block_propagation_sleep_duration_secs {
            config.block_propagation_sleep_duration =
                Duration::from_secs(block_propagation_sleep_duration);
        }
        if let Some(recoverable_error_sleep_duration) = self.recoverable_error_sleep_duration_secs {
            config.recoverable_error_sleep_duration =
                Duration::from_secs(recoverable_error_sleep_duration);
        }
    }
}
