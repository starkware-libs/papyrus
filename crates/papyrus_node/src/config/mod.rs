#[cfg(test)]
mod config_test;

mod file_config;

use std::collections::HashMap;
use std::mem::discriminant;
use std::path::PathBuf;
use std::time::Duration;
use std::{env, fs, io};

use clap::{arg, value_parser, Arg, ArgMatches, Command};
use file_config::FileConfigFormat;
use papyrus_gateway::GatewayConfig;
use papyrus_monitoring_gateway::MonitoringGatewayConfig;
use papyrus_storage::db::DbConfig;
use papyrus_storage::StorageConfig;
use papyrus_sync::{CentralSourceConfig, SyncConfig};
use serde::{Deserialize, Serialize};
use starknet_api::core::ChainId;
use starknet_client::RetryConfig;

// The path of the default configuration file, provided as part of the crate.
const CONFIG_FILE: &str = "config/config.yaml";

/// The configurations of the various components of the node.
#[derive(Deserialize, Serialize, Clone)]
pub struct Config {
    pub gateway: GatewayConfig,
    pub central: CentralSourceConfig,
    pub monitoring_gateway: MonitoringGatewayConfig,
    pub storage: StorageConfig,
    /// None if the syncing should be disabled.
    pub sync: Option<SyncConfig>,
}

impl Config {
    pub fn load(args: Vec<String>) -> Result<Self, ConfigError> {
        ConfigBuilder::build(args)
    }

    pub fn get_config_representation(&self) -> Result<serde_yaml::Value, ConfigError> {
        Ok(serde_yaml::to_value(FileConfigFormat::from(self.clone()))?)
    }
}

#[derive(thiserror::Error, Debug)]
pub enum ConfigError {
    #[error("Unable to parse path: {path}")]
    BadPath { path: PathBuf },
    #[error(transparent)]
    Clap(#[from] clap::Error),
    #[error(transparent)]
    Matches(#[from] clap::parser::MatchesError),
    #[error(transparent)]
    Read(#[from] io::Error),
    #[error(transparent)]
    Serde(#[from] serde_yaml::Error),
    #[error(
        "CLA http_header \"{illegal_header}\" is not valid. The Expected format is name:value"
    )]
    CLAHttpHeader { illegal_header: String },
}

// Builds the configuration for the node based on default values, yaml configuration file and
// command-line arguments.
// TODO: add configuration from env variables.
pub(crate) struct ConfigBuilder {
    args: Option<ArgMatches>,
    chain_id: ChainId,
    config: Config,
}

// Default configuration values.
// TODO: Consider implementing Default for each component individually.
impl Default for ConfigBuilder {
    fn default() -> Self {
        let chain_id = ChainId(String::from("SN_MAIN"));

        ConfigBuilder {
            args: None,
            chain_id: chain_id.clone(),
            config: Config {
                central: CentralSourceConfig {
                    concurrent_requests: 300,
                    url: String::from("https://alpha-mainnet.starknet.io/"),
                    http_headers: None,
                    retry_config: RetryConfig {
                        retry_base_millis: 30,
                        retry_max_delay_millis: 30000,
                        max_retries: 10,
                    },
                },
                gateway: GatewayConfig {
                    chain_id,
                    server_address: String::from("0.0.0.0:8080"),
                    max_events_chunk_size: 1000,
                    max_events_keys: 100,
                },
                monitoring_gateway: MonitoringGatewayConfig {
                    server_address: String::from("0.0.0.0:8081"),
                },
                storage: StorageConfig {
                    db_config: DbConfig { path: String::from("./data"), max_size: 1099511627776 },
                },
                sync: Some(SyncConfig {
                    block_propagation_sleep_duration: Duration::from_secs(10),
                    recoverable_error_sleep_duration: Duration::from_secs(10),
                }),
            },
        }
    }
}

impl ConfigBuilder {
    // Creates the configuration struct.
    fn build(args: Vec<String>) -> Result<Config, ConfigError> {
        Ok(Self::default().prepare_command(args)?.yaml()?.args()?.propagate_chain_id().config)
    }

    // Builds the applications command-line interface.
    fn prepare_command(mut self, args: Vec<String>) -> Result<Self, ConfigError> {
        self.args = Some(
            Command::new("Papyrus",)
            .version("Pre-release")
            .about("Papyrus is a StarkNet full node written in Rust.")
            .args(&[
                arg!(-f --config_file [path] "Optionally sets a config file to use").value_parser(value_parser!(PathBuf)),
                arg!(-c --chain_id [name] "Optionally sets chain id to use"),
                arg!(--server_address ["IP:PORT"] "Optionally sets the RPC listening address"),
                arg!(--http_headers ["NAME:VALUE"] ... "Optionally adds headers to the http requests"),
                arg!(-s --storage [path] "Optionally sets storage path to use (automatically extended with chain ID)").value_parser(value_parser!(PathBuf)),
                arg!(-n --no_sync [bool] "Optionally run without sync").value_parser(value_parser!(bool)).default_missing_value("true"),
                arg!(--central_url ["URL"] "Central URL. It should match chain_id."),
            ])
            .try_get_matches_from(args).unwrap_or_else(|e| e.exit()),
        );
        Ok(self)
    }

    // Parses a yaml configuration file given by the command-line args (or default), and applies it
    // on the configuration.
    fn yaml(mut self) -> Result<Self, ConfigError> {
        let mut yaml_path = CONFIG_FILE;

        let args = self.args.clone().expect("Config builder should have args.");
        if let Some(config_file) = args.try_get_one::<PathBuf>("config_file")? {
            yaml_path =
                config_file.to_str().ok_or(ConfigError::BadPath { path: config_file.clone() })?;
        }

        let yaml_contents = fs::read_to_string(yaml_path)?;
        let from_yaml: FileConfigFormat = serde_yaml::from_str(&yaml_contents)?;
        from_yaml.update_config(&mut self);

        Ok(self)
    }

    // Reads the command-line args and updates the relevant configurations.
    fn args(mut self) -> Result<Self, ConfigError> {
        match self.args {
            None => unreachable!(),
            Some(ref args) => {
                if let Some(chain_id) = args.try_get_one::<String>("chain_id")? {
                    self.chain_id = ChainId(chain_id.clone());
                }

                if let Some(server_address) = args.try_get_one::<String>("server_address")? {
                    self.config.gateway.server_address = server_address.to_string()
                }

                if let Some(storage_path) = args.try_get_one::<PathBuf>("storage")? {
                    self.config.storage.db_config.path = storage_path
                        .to_str()
                        .ok_or(ConfigError::BadPath { path: storage_path.clone() })?
                        .to_owned();
                }

                if let Some(http_headers) = args.try_get_one::<String>("http_headers")? {
                    let mut headers_map = match self.config.central.http_headers {
                        Some(map) => map,
                        None => HashMap::new(),
                    };
                    for header in http_headers.split(' ') {
                        let split: Vec<&str> = header.split(':').collect();
                        if split.len() != 2 {
                            return Err(ConfigError::CLAHttpHeader {
                                illegal_header: header.to_string(),
                            });
                        }
                        headers_map.insert(split[0].to_string(), split[1].to_string());
                    }
                    self.config.central.http_headers = Some(headers_map);
                }

                if let Some(no_sync) = args.try_get_one::<bool>("no_sync")? {
                    if *no_sync {
                        self.config.sync = None;
                    }
                }
                if let Some(central_url) = args.try_get_one::<String>("central_url")? {
                    self.config.central.url = central_url.to_string()
                }

                Ok(self)
            }
        }
    }

    // Propagates the chain id into all the of configurations that use it.
    fn propagate_chain_id(mut self) -> Self {
        self.config.gateway.chain_id = self.chain_id.clone();
        // Assuming a valid path.
        self.config.storage.db_config.path.push_str(format!("/{}", self.chain_id.0).as_str());
        self
    }
}
