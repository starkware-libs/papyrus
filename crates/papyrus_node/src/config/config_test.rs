use std::collections::HashMap;
use std::env::{self, args};
use std::io::Write;
use std::path::{Path, PathBuf};

use starknet_api::core::ChainId;
use tempfile::NamedTempFile;
use test_utils::get_absolute_path;

use crate::config::{Config, ConfigBuilder};

#[test]
fn load_default_config() {
    env::set_current_dir(get_absolute_path("")).expect("Couldn't set working dir.");
    // TODO(spapini): Move the config closer.
    Config::load(vec![]).expect("Failed to load the config.");
}

#[test]
fn default_builder() {
    let builder = ConfigBuilder::default();
    assert_eq!(builder.config.gateway.chain_id, ChainId("SN_MAIN".to_owned()));
    assert!(builder.config.sync.is_some())
}

#[test]
fn prepare_command() {
    let args = vec![
        "Papyrus".to_owned(),
        "--config_file=conf.yaml".to_owned(),
        "--chain_id=CHAIN_ID".to_owned(),
        "--server_address=IP:PORT".to_owned(),
        "--http_headers=NAME_1:VALUE_1 NAME_2:VALUE_2".to_owned(),
        "--storage=path".to_owned(),
        "--no_sync=true".to_owned(),
        "--central_url=URL".to_owned(),
    ];
    let builder = ConfigBuilder::default().prepare_command(args).unwrap();
    let builder_args = builder.args.expect("Expected to have args");

    assert_eq!(
        *builder_args.get_one::<PathBuf>("config_file").expect("Expected to have config arg"),
        PathBuf::from("conf.yaml")
    );
    assert_eq!(
        *builder_args.get_one::<String>("chain_id").expect("Expected to have chain_id arg"),
        "CHAIN_ID".to_owned()
    );
    assert_eq!(
        *builder_args
            .get_one::<String>("server_address")
            .expect("Expected to have server_address arg"),
        "IP:PORT".to_owned()
    );

    let headers_list: Vec<&str> = builder_args
        .get_one::<String>("http_headers")
        .expect("Expected to have http_headers args")
        .split(' ')
        .collect();
    assert_eq!(headers_list.len(), 2);
    assert_eq!(headers_list[0].to_owned(), "NAME_1:VALUE_1".to_owned());
    assert_eq!(headers_list[1].to_owned(), "NAME_2:VALUE_2".to_owned());

    assert_eq!(
        *builder_args.get_one::<PathBuf>("storage").expect("Expected to have storage arg"),
        PathBuf::from("path")
    );
    let no_sync = *builder_args.get_one::<bool>("no_sync").expect("Expected to have no_sync arg");
    assert!(no_sync);
    assert_eq!(
        *builder_args.get_one::<String>("central_url").expect("Expected to have central_url arg"),
        "URL".to_owned()
    );
}

#[test]
fn load_yaml_config() {
    let mut f = NamedTempFile::new().unwrap();
    let yaml = r"
chain_id: TEST
gateway:
    max_events_keys: 1234
";
    f.write_all(yaml.as_bytes()).unwrap();
    let args = vec!["Papyrus".to_owned(), format!("--config_file={}", f.path().to_str().unwrap())];
    let builder = ConfigBuilder::default().prepare_command(args).unwrap().yaml().unwrap();

    assert_eq!(builder.chain_id, ChainId("TEST".to_owned()));
    assert_eq!(builder.config.gateway.max_events_keys, 1234);
}

#[test]
fn load_http_headers() {
    let mut f = NamedTempFile::new().unwrap();
    let yaml = r"
central:
    http_headers:
        NAME_1: VALUE_1
        NAME_2: VALUE_2 
";
    f.write_all(yaml.as_bytes()).unwrap();
    let args = vec![
        "Papyrus".to_owned(),
        format!("--config_file={}", f.path().to_str().unwrap()),
        "--http_headers=NAME_2:NEW_VALUE_2 NAME_3:VALUE_3".to_owned(),
    ];
    let builder =
        ConfigBuilder::default().prepare_command(args).unwrap().yaml().unwrap().args().unwrap();

    let target_http_headers = HashMap::from([
        ("NAME_1".to_string(), "VALUE_1".to_string()),
        ("NAME_2".to_string(), "NEW_VALUE_2".to_string()),
        ("NAME_3".to_string(), "VALUE_3".to_string()),
    ]);
    assert_eq!(builder.config.central.http_headers.unwrap(), target_http_headers);
}
