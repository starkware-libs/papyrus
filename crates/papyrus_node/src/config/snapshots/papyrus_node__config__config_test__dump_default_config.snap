---
source: crates/papyrus_node/src/config/config_test.rs
expression: dumped_default_config
---
{
  "base_layer.node_url": {
    "description": "A required param! Ethereum node URL. A schema to match to Infura node: https://mainnet.infura.io/v3/<your_api_key>, but any other node can be used.",
    "param_type": "String",
    "privacy": "Private"
  },
  "base_layer.starknet_contract_address": {
    "description": "Starknet contract address in ethereum.",
    "value": "0xc662c410C0ECf747543f5bA90660f6ABeBD9C8c4",
    "privacy": "Public"
  },
  "central.class_cache_size": {
    "description": "Size of class cache, must be a positive integer.",
    "value": {
      "$serde_json::private::Number": "100"
    },
    "privacy": "Public"
  },
  "central.concurrent_requests": {
    "description": "Maximum number of concurrent requests to Starknet feeder-gateway for getting a type of data (for example, blocks).",
    "value": {
      "$serde_json::private::Number": "10"
    },
    "privacy": "Public"
  },
  "central.http_headers": {
    "description": "'k1:v1 k2:v2 ...' headers for SN-client.",
    "value": "",
    "privacy": "Private"
  },
  "central.max_classes_to_download": {
    "description": "Maximum number of classes to download at a given time.",
    "value": {
      "$serde_json::private::Number": "20"
    },
    "privacy": "Public"
  },
  "central.max_state_updates_to_download": {
    "description": "Maximum number of state updates to download at a given time.",
    "value": {
      "$serde_json::private::Number": "20"
    },
    "privacy": "Public"
  },
  "central.max_state_updates_to_store_in_memory": {
    "description": "Maximum number of state updates to store in memory at a given time.",
    "value": {
      "$serde_json::private::Number": "20"
    },
    "privacy": "Public"
  },
  "central.retry_config.max_retries": {
    "description": "Maximum number of retries before the node stops retrying.",
    "value": {
      "$serde_json::private::Number": "10"
    },
    "privacy": "Public"
  },
  "central.retry_config.retry_base_millis": {
    "description": "Base waiting time after a failed request. After that, the time increases exponentially.",
    "value": {
      "$serde_json::private::Number": "30"
    },
    "privacy": "Public"
  },
  "central.retry_config.retry_max_delay_millis": {
    "description": "Max waiting time after a failed request.",
    "value": {
      "$serde_json::private::Number": "30000"
    },
    "privacy": "Public"
  },
  "central.url": {
    "description": "Starknet feeder-gateway URL. It should match chain_id.",
    "value": "https://alpha-mainnet.starknet.io/",
    "privacy": "Public"
  },
  "monitoring_gateway.collect_metrics": {
    "description": "If true, collect and return metrics in the monitoring gateway.",
    "value": false,
    "privacy": "Public"
  },
  "monitoring_gateway.metric_labels": {
    "description": "'label1:value1 label2:value2 ...' additional labels for metrics.",
    "value": "",
    "privacy": "Public"
  },
  "monitoring_gateway.present_full_config_secret": {
    "description": "A secret for presenting the full general config. If no value is provided, the system will generate one.",
    "param_type": "String",
    "privacy": "Private"
  },
  "monitoring_gateway.server_address": {
    "description": "node's monitoring server.",
    "value": "0.0.0.0:8081",
    "privacy": "Public"
  },
  "monitoring_gateway.starknet_url": {
    "description": "The URL of a centralized Starknet gateway.",
    "value": "https://alpha-mainnet.starknet.io/",
    "privacy": "Public"
  },
  "network.#is_none": {
    "description": "Flag for an optional field",
    "value": true,
    "privacy": "TemporaryValue"
  },
  "network.header_buffer_size": {
    "description": "Size of the buffer for headers read from the storage.",
    "value": {
      "$serde_json::private::Number": "100000"
    },
    "privacy": "Public"
  },
  "network.idle_connection_timeout": {
    "description": "Amount of time in seconds that a connection with no active sessions will stay alive.",
    "value": {
      "$serde_json::private::Number": "10"
    },
    "privacy": "Public"
  },
  "network.peer.#is_none": {
    "description": "Flag for an optional field",
    "value": true,
    "privacy": "TemporaryValue"
  },
  "network.peer.ip": {
    "description": "The ipv4 address of another peer that the node will dial to.",
    "value": "127.0.0.1",
    "privacy": "Public"
  },
  "network.peer.peer_id": {
    "description": "Peer ID to send requests to. If not set, the node will not send requests. for info: https://docs.libp2p.io/concepts/fundamentals/peers/",
    "value": "QmYyQSo1c1Ym7orWxLYvCrM2EmxFTANf8wXmmE7DWjhx5N",
    "privacy": "Public"
  },
  "network.peer.tcp_port": {
    "description": "The port on the other peer that the node will dial to to use for TCP transport.",
    "value": {
      "$serde_json::private::Number": "10002"
    },
    "privacy": "Public"
  },
  "network.quic_port": {
    "description": "The port that the node listens on for incoming quic connections.",
    "value": {
      "$serde_json::private::Number": "10001"
    },
    "privacy": "Public"
  },
  "network.session_timeout": {
    "description": "Maximal time in seconds that each session can take before failing on timeout.",
    "value": {
      "$serde_json::private::Number": "10"
    },
    "privacy": "Public"
  },
  "network.tcp_port": {
    "description": "The port that the node listens on for incoming tcp connections.",
    "value": {
      "$serde_json::private::Number": "10000"
    },
    "privacy": "Public"
  },
  "p2p_sync.#is_none": {
    "description": "Flag for an optional field",
    "value": true,
    "privacy": "TemporaryValue"
  },
  "p2p_sync.num_headers_per_query": {
    "description": "The maximum amount of headers to ask from peers in each iteration.",
    "value": {
      "$serde_json::private::Number": "10000"
    },
    "privacy": "Public"
  },
  "p2p_sync.query_timeout": {
    "description": "Time in seconds to wait for query responses until we mark it as failed",
    "value": {
      "$serde_json::private::Number": "5"
    },
    "privacy": "Public"
  },
  "rpc.chain_id": {
    "description": "The chain to follow. For more details see https://docs.starknet.io/documentation/architecture_and_concepts/Blocks/transactions/#chain-id.",
    "value": "SN_MAIN",
    "privacy": "Public"
  },
  "rpc.collect_metrics": {
    "description": "If true, collect metrics for the rpc.",
    "value": false,
    "privacy": "Public"
  },
  "rpc.execution_config": {
    "description": "Path to the execution configuration file.",
    "value": "config/execution/mainnet.json",
    "privacy": "Public"
  },
  "rpc.max_events_chunk_size": {
    "description": "Maximum chunk size supported by the node in get_events requests.",
    "value": {
      "$serde_json::private::Number": "1000"
    },
    "privacy": "Public"
  },
  "rpc.max_events_keys": {
    "description": "Maximum number of keys supported by the node in get_events requests.",
    "value": {
      "$serde_json::private::Number": "100"
    },
    "privacy": "Public"
  },
  "rpc.server_address": {
    "description": "IP:PORT of the node`s JSON-RPC server.",
    "value": "0.0.0.0:8080",
    "privacy": "Public"
  },
  "rpc.starknet_gateway_retry_config.max_retries": {
    "description": "For communicating with Starknet gateway, maximum number of retries before the node stops retrying.",
    "value": {
      "$serde_json::private::Number": "5"
    },
    "privacy": "Public"
  },
  "rpc.starknet_gateway_retry_config.retry_base_millis": {
    "description": "For communicating with Starknet gateway, base waiting time after a failed request. After that, the time increases exponentially.",
    "value": {
      "$serde_json::private::Number": "50"
    },
    "privacy": "Public"
  },
  "rpc.starknet_gateway_retry_config.retry_max_delay_millis": {
    "description": "For communicating with Starknet gateway, max waiting time after a failed request.",
    "value": {
      "$serde_json::private::Number": "1000"
    },
    "privacy": "Public"
  },
  "rpc.starknet_url": {
    "description": "URL for communicating with Starknet in write_api methods.",
    "value": "https://alpha-mainnet.starknet.io/",
    "privacy": "Public"
  },
  "storage.db_config.chain_id": {
    "description": "The chain to follow. For more details see https://docs.starknet.io/documentation/architecture_and_concepts/Blocks/transactions/#chain-id.",
    "value": "SN_MAIN",
    "privacy": "Public"
  },
  "storage.db_config.enforce_file_exists": {
    "description": "Whether to enforce that the path exists. If true, `open_env` fails when the mdbx.dat file does not exist.",
    "value": false,
    "privacy": "Public"
  },
  "storage.db_config.growth_step": {
    "description": "The growth step in bytes, must be greater than zero to allow the database to grow.",
    "value": {
      "$serde_json::private::Number": "4294967296"
    },
    "privacy": "Public"
  },
  "storage.db_config.max_size": {
    "description": "The maximum size of the node's storage in bytes.",
    "value": {
      "$serde_json::private::Number": "1099511627776"
    },
    "privacy": "Public"
  },
  "storage.db_config.min_size": {
    "description": "The minimum size of the node's storage in bytes.",
    "value": {
      "$serde_json::private::Number": "1048576"
    },
    "privacy": "Public"
  },
  "storage.db_config.path_prefix": {
    "description": "Prefix of the path of the node's storage directory, the storage file path will be <path_prefix>/<chain_id>. The path is not created automatically.",
    "value": "./data",
    "privacy": "Public"
  },
  "storage.mmap_file_config.growth_step": {
    "description": "The growth step in bytes, must be greater than max_object_size.",
    "value": {
      "$serde_json::private::Number": "1073741824"
    },
    "privacy": "Public"
  },
  "storage.mmap_file_config.max_object_size": {
    "description": "The maximum size of a single object in the file in bytes",
    "value": {
      "$serde_json::private::Number": "1048576"
    },
    "privacy": "Public"
  },
  "storage.mmap_file_config.max_size": {
    "description": "The maximum size of a memory mapped file in bytes. Must be greater than growth_step.",
    "value": {
      "$serde_json::private::Number": "1099511627776"
    },
    "privacy": "Public"
  },
  "storage.scope": {
    "description": "The categories of data saved in storage.",
    "value": "FullArchive",
    "privacy": "Public"
  },
  "sync.#is_none": {
    "description": "Flag for an optional field",
    "value": false,
    "privacy": "TemporaryValue"
  },
  "sync.base_layer_propagation_sleep_duration": {
    "description": "Time in seconds to poll the base layer to get the latest proved block.",
    "value": {
      "$serde_json::private::Number": "10"
    },
    "privacy": "Public"
  },
  "sync.block_propagation_sleep_duration": {
    "description": "Time in seconds before checking for a new block after the node is synchronized.",
    "value": {
      "$serde_json::private::Number": "2"
    },
    "privacy": "Public"
  },
  "sync.blocks_max_stream_size": {
    "description": "Max amount of blocks to download in a stream.",
    "value": {
      "$serde_json::private::Number": "1000"
    },
    "privacy": "Public"
  },
  "sync.recoverable_error_sleep_duration": {
    "description": "Waiting time in seconds before restarting synchronization after a recoverable error.",
    "value": {
      "$serde_json::private::Number": "3"
    },
    "privacy": "Public"
  },
  "sync.state_updates_max_stream_size": {
    "description": "Max amount of state updates to download in a stream.",
    "value": {
      "$serde_json::private::Number": "1000"
    },
    "privacy": "Public"
  },
  "sync.verify_blocks": {
    "description": "Whether to verify incoming blocks.",
    "value": true,
    "privacy": "Public"
  }
}
